### ISSUE TITLE
False Positive [E0308] mismatched types in returned types of surrealdb crate

### ISSUE THREAD

**Issue by Maksim20023** [intellij-rust/intellij-rust#10796]
## Environment

* **IntelliJ Rust plugin version:** 0.4.199.5414-231 / 0.4.199.5415-232 / 0.4.200.5421-232
* **Rust toolchain version:** 1.71.1
* **IDE name and version:** CL-231.9225.21 / CL - 232.8660.186
* **Operating system:** OS Windows 10(amd64)

## Problem description

The following code calls Surrealdb's Client library. They have an id type called Thing to identify a specific record. This makes the return type always resolve to Option. The analyzer however resolves the type to Vec, which is the case for non Thing ids. The feedback is a hard error, but the code is correct. The struct (Self/Saga) has an id of type Thing.

![image-2](https://github.com/intellij-rust/intellij-rust/assets/135699123/a0d043cc-6292-49d5-bb3a-deeb1cac1091)

## Steps to reproduce

You can just set up a call to surrealdb:s update call with a
Thing (their own id t as the id.

A short summary of the case:

Cargo.toml:
````toml
surrealdb = "1.0.0-beta.9+20230402"
````
To get a Client set up something like this:

````rust
/// Create database connection pool for persistence
pub async fn get_db_pool() -> Result<Surreal<Client>> {
    // --- Connect to the server
    let client = Surreal::new::<Ws>(
        &*format!("{}:{}",
                  env::var("SURREAL_HOST").expect("SURREAL_HOST not in env
vars"),
                  env::var("SURREAL_PORT").expect("SURREAL_PORT not in env
vars"),
        )).with_capacity(5).await?;

    // --- Sign in as root
    client.signin(Root {
        username: &env::var("SURREAL_USER").expect("SURREAL_USER not in env
vars"),
        password: &env::var("SURREAL_PASSWORD").expect("SURREAL_PASSWORD
not in env vars"),
    }).await?;

    client.use_ns("test").use_db(&env::var("SURREAL_DB").expect("SURREAL_DB
not in env vars")).await?;

    Ok(client)
}
````
If you want a full runnable example to make sure it is in fact the correct
behaviour you can run the db with:

The update:

````rust
struct Test {
    id: Thing,
    data: i32,
}
impl Test {
    pub async fn update(&self, db: Surreal<Client>) -> Result<()> {
        let _: Option<Self> =
db.update(self.id.clone()).content(&self).await?;
        Ok(())
    }
}
````

If you want a full runnable example to make sure it is in fact the correct
behaviour you can run the db with:
````rust
docker run -d -it --rm --name surrealdb -p 8000:8000
surrealdb/surrealdb:latest start --log trace --user testuser --pass test
memory
````
And then of course just create some driver code to create a client and call
update.
